### PR TITLE
[DDO-2831] Yale: add support for using Cloud Monitoring metrics to identify key last auth time

### PR DESCRIPTION
**Pull request by choover-broad**
Add a library to Yale that supports using Cloud Monitoring metrics instead of Policy Analyzer to determine when a key was last used to authenticate. This gets us data that is at most 1-6 hours old, instead of 2-9 days.

### Testing

This PR includes a unit test that utilizes the [go-replayers](https://github.com/google/go-replayers) library. I chose this because I wanted to experiment with replayers to avoid the tedious work of generating and setting up mocks. Time will tell how high-maintenance it is. One thing I do like is that the tests run once, locally, against real APIs, which is useful for verifying mocks are correct.

### Risk Assessment

I consider this PR low risk because it does not meaningfully change Yale's behavior, only adds a new library to the codebase.

### Background

Currently, Yale uses the Policy Analyzer API to determine if a service account key is still in use and therefore, unsafe to disable. Unfortunately, the PolicyAnalyzer's data can lag a variable amount of time (between 2-9 days in my experience) behind realtime, meaning that the data it returns is not super useful; we have to make Yale extremely conservative about rotation in order to successfully use it.

This PR adds support for using the `iam.googleapis.com/service_account/key/authn_events_count` Cloud Monitoring metric to determine if it is safe to disable a key. Based on experiments over the past few days, the data returned by this metric lags at most 2-6 hours behind realtime. I expect this window to remain short because of the different end-user use cases for Cloud Monitoring and Policy Analyzer.

Note that this PR does *not* update Yale to use Cloud Monitoring, it merely adds a new library with test coverage. Yale will be cut over to using the new `authmetrics` library in a future PR.

### Example Policy Analyzer API call

Here's an example response that demonstrates the issue:
```
# gcloud policy-intelligence query-activity \
  --activity-type=serviceAccountKeyLastAuthentication \
  --project=broad-dsde-dev \
  --query-filter='activities.full_resource_name="//iam.googleapis.com/projects/broad-dsde-dev/serviceAccounts/agora-dev-service-account@broad-dsde-dev.iam.gserviceaccount.com/keys/6a1ab7088e1690094952b6857f04bc294b61b03e"'
 
---
activity:
  lastAuthenticatedTime: '2023-02-15T08:00:00Z'
  serviceAccountKey:
    fullResourceName: //iam.googleapis.com/projects/broad-dsde-dev/serviceAccounts/agora-dev-service-account@broad-dsde-dev.iam.gserviceaccount.com/keys/6a1ab7088e1690094952b6857f04bc294b61b03e
    projectNumber: '806222273987'
    serviceAccountId: '102047007757533412597'
activityType: serviceAccountKeyLastAuthentication
fullResourceName: //iam.googleapis.com/projects/broad-dsde-dev/serviceAccounts/agora-dev-service-account@broad-dsde-dev.iam.gserviceaccount.com/keys/6a1ab7088e1690094952b6857f04bc294b61b03e
observationPeriod:
  endTime: '2023-05-04T07:00:00Z'
  startTime: '2021-05-24T07:00:00Z'
```

At the time of this writing, the observation period ends four days ago; but based on past experience, it can lag up to 8 days behind current time.

### Example Cloud Monitoring Metrics Data

[Here](https://console.cloud.google.com/monitoring/metrics-explorer;duration=PT3H?pageState=%7B%22domainObjectDeprecationId%22:%2248E8526C-9F20-4639-9B14-7535454B8EB6%22,%22title%22:%22Untitled%22,%22xyChart%22:%7B%22constantLines%22:%5B%5D,%22dataSets%22:%5B%7B%22plotType%22:%22LINE%22,%22targetAxis%22:%22Y1%22,%22timeSeriesFilter%22:%7B%22aggregations%22:%5B%7B%22crossSeriesReducer%22:%22REDUCE_NONE%22,%22groupByFields%22:%5B%5D,%22perSeriesAligner%22:%22ALIGN_RATE%22%7D,%7B%22crossSeriesReducer%22:%22REDUCE_MEAN%22,%22groupByFields%22:%5B%22metric.label.%5C%22key_id%5C%22%22%5D,%22perSeriesAligner%22:%22ALIGN_MEAN%22%7D%5D,%22apiSource%22:%22DEFAULT_CLOUD%22,%22crossSeriesReducer%22:%22REDUCE_NONE%22,%22filter%22:%22metric.type%3D%5C%22iam.googleapis.com%2Fservice_account%2Fkey%2Fauthn_events_count%5C%22%20resource.type%3D%5C%22iam_service_account%5C%22%22,%22groupByFields%22:%5B%5D,%22minAlignmentPeriod%22:%2260s%22,%22perSeriesAligner%22:%22ALIGN_RATE%22,%22secondaryCrossSeriesReducer%22:%22REDUCE_MEAN%22,%22secondaryGroupByFields%22:%5B%22metric.label.%5C%22key_id%5C%22%22%5D%7D%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22y1Axis%22:%7B%22label%22:%22%22,%22scale%22:%22LINEAR%22%7D%7D%7D&project=broad-dsde-dev) is an example graph showing data from the `iam.googleapis.com/service_account/key/authn_events_count` metric, lagging 30 minutes behind current.
Screen Shot 2023-05-08 at 10 44 37 AM

### CODE REVIEW

File: internal/yale/authmetrics/authmetrics.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		keyId, exists := resp.GetMetric().GetLabels()[keyIdLabel]
+		if !exists {
+			return nil, fmt.Errorf("time series %s missing label: %s", metricType, keyIdLabel)
+		}
+		serviceAccountId, exists := resp.GetResource().Labels[uniqueIdLabel]
+		if !exists {
+			return nil, fmt.Errorf("time series %s missing resource label: %s", metricType, uniqueIdLabel)
+		}
+		serviceAccountEmail, exists := serviceAccountIds[serviceAccountId]
+		if !exists {
```
Comment (mflinn-broad) [inline internal/yale/authmetrics/authmetrics.go:123]: Appreciate the thorough handling of all the edge cases calling the metrics API here.

**Review by jack-r-warren — APPROVED** [review]
LGTM

**Review by mflinn-broad — COMMENTED** [review]
Lgtm, replayer is interesting will have to try it out.

**Review by mflinn-broad — APPROVED** [review]
whoops meant to approve

**Pull request merged**